### PR TITLE
KLL doubles

### PR DESCRIPTION
**Pull request by leerho**
This adds a KllDoublesSketch to the library. All the tests work including the detailed verification test Keven wrote that checks the internal arrays at multiple stages of N. We probably should run some characterization tests before we release this.

### CODE REVIEW

File: src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+import org.apache.datasketches.SketchesArgumentException;
+
+/**
+ * Static methods to support KllSketch
+ * @author Kevin Lang
+ * @author Alexander Saydakov
+ */
+public class KllDoublesHelper {
+
+  static boolean isEven(final int value) {
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java:36]: perhaps, such methods, identical in both KllFloatsHelper and KllDoublesHepler, should be shared in common helper, say, KllHelper?

File: src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    return (value & 1) == 0;
+  }
+
+  static boolean isOdd(final int value) {
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java:40]: also identical to one in KllFloatsHelper

File: src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+  /**
+   * Copy the old array into a new larger array.
+   * The extra space is at the top.
+   * @param oldArr the given old array with data
+   * @param newLen the new length larger than the oldArr.length.
+   * @return the new array
+   */
+  static int[] growIntArray(final int[] oldArr, final int newLen) {
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java:68]: also identical

File: src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+  /**
+   * Returns the maximum number of items that this sketch can handle
+   * @param k The sizing / accuracy parameter of the sketch in items.
+   * Note: this method actually works for k values up to k = 2^29 and 61 levels,
+   * however only k values up to (2^16 - 1) are currently used by the sketch.
+   * @param m the size of the smallest level in items.
+   * @param numLevels the upper bound number of levels based on <i>n</i> items.
+   * @return the total item capacity of the sketch.
+   */
+  static int computeTotalCapacity(final int k, final int m, final int numLevels) {
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java:94]: also a duplicate

File: src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+  /**
+   * Returns the capacity of a specific level.
+   * @param k the accuracy parameter of the sketch. Maximum is 2^29.
+   * @param numLevels the number of current levels in the sketch. Maximum is 61.
+   * @param height the zero-based index of a level with respect to the smallest level.
+   * This varies from 0 to 60.
+   * @param minWidth the minimum level width. Default is 8.
+   * @return the capacity of a specific level
+   */
+  static int levelCapacity(final int k, final int numLevels, final int height, final int minWidth) {
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java:111]: duplicate

File: src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    return intCapAuxAux(tmp, rest);
+  }
+
+  /**
+   * Performs the integer based calculation of an individual level (or folded level).
+   * @param k the configured k of the sketch
+   * @param depth depth the zero-based index of the level being computed.
+   * @return the actual capacity of a given level given its depth index.
+   */
+  private static long intCapAuxAux(final long k, final int depth) {
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java:142]: duplicate

File: src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+   * This is the exact powers of 3 from 3^0 to 3^30 where the exponent is the index
+   */
+  private static final long[] powersOfThree =
+      new long[] {1, 3, 9, 27, 81, 243, 729, 2187, 6561, 19683, 59049, 177147, 531441,
+  1594323, 4782969, 14348907, 43046721, 129140163, 387420489, 1162261467,
+  3486784401L, 10460353203L, 31381059609L, 94143178827L, 282429536481L,
+  847288609443L, 2541865828329L, 7625597484987L, 22876792454961L, 68630377364883L,
+  205891132094649L};
+
+  static long sumTheSampleWeights(final int num_levels, final int[] levels) {
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/kll/KllDoublesHelper.java:160]: duplicate

**Comment by leerho** [comment]
Fair enough.  I will rejigger the methods to leverage the common ones, then refresh this PR

**Review by AlexanderSaydakov — APPROVED** [review]

**Pull request merged**